File: Sources/PimTrayCore/Providers/AzureResourceProvider.swift
import Foundation

/// PIM for Azure resource roles through Azure Resource Manager.
public struct AzureResourceProvider: PIMProvider {
    public let kind: RoleScopeKind = .azureResource
    public let scopes = ArmScopes.all
    let transport: GraphTransport

    public init(http: any HTTPClient, tokens: any TokenProviding) {
        transport = GraphTransport(http: http, tokens: tokens, mapper: GraphTransport.mapArmError)
    }

    // MARK: Wire models

    struct Named: Decodable { let displayName: String?; let type: String?; let id: String? }
    struct Expanded: Decodable { let scope: Named?; let roleDefinition: Named? }
    struct Expiration: Decodable { let type: String?; let duration: String?; let endDateTime: Date? }
    struct ScheduleInfo: Decodable { let startDateTime: Date?; let expiration: Expiration? }
    struct Properties: Decodable {
        let scope: String
        let roleDefinitionId: String
        let principalId: String?
        let status: String?
        let assignmentType: String?
        let roleEligibilityScheduleId: String?
        let linkedRoleEligibilityScheduleId: String?
        let startDateTime: Date?
        let endDateTime: Date?
        let createdOn: Date?
        let scheduleInfo: ScheduleInfo?
        let expandedProperties: Expanded?
    }
    struct Instance: Decodable { let name: String; let id: String; let properties: Properties }
    struct Page<T: Decodable>: Decodable { let value: [T]; let nextLink: String? }

    static let pendingStatuses: Set<String> = ["PendingApproval", "PendingAdminDecision", "PendingApprovalProvisioning"]

    func armURL(_ path: String, apiVersion: String = "2020-10-01", query: [String: String] = [:]) throws -> URL {
        guard var components = URLComponents(url: GraphTransport.armBase.appendingPathComponent(path), resolvingAgainstBaseURL: false) else {
            throw PIMError.unexpected(status: 0, body: "Bad ARM path \(path)")
        }
        var items = [URLQueryItem(name: "api-version", value: apiVersion)]
        for (k, v) in query.sorted(by: { $0.key < $1.key }) { items.append(URLQueryItem(name: k, value: v)) }
        components.queryItems = items
        guard let url = components.url else { throw PIMError.unexpected(status: 0, body: "Bad ARM URL \(path)") }
        return url
    }

    /// GET every page of an ARM list, following `nextLink`.
    func listAll<T: Decodable>(_ type: T.Type, identity: Identity, tenantId: String, url: URL) async throws -> [T] {
        var next: URL? = url
        var out: [T] = []
        while let current = next {
            let r = try await transport.get(identity: identity, tenantId: tenantId, url: current, scopes: scopes)
            let page = try GraphJSON.decoder.decode(Page<T>.self, from: r.body)
            out += page.value
            next = page.nextLink.flatMap(URL.init(string:))
        }
        return out
    }

    static func caption(_ e: Expanded?) -> String? {
        guard let scope = e?.scope, let name = scope.displayName else { return nil }
        let type: String?
        switch scope.type?.lowercased() {
        case "subscription": type = "subscription"
        case "resourcegroup": type = "resource group"
        case "managementgroup": type = "management group"
        case let other?: type = other
        case nil: type = nil
        }
        return type.map { "\(name) · \($0)" } ?? name
    }

    // MARK: Reads

    public func eligibleRoles(identity: Identity, tenant: TenantContext) async throws -> [EligibleRole] {
        let url = try armURL("providers/Microsoft.Authorization/roleEligibilityScheduleInstances", query: ["$filter": "asTarget()"])
        let items = try await listAll(Instance.self, identity: identity, tenantId: tenant.tenantId, url: url)
        var seen = Set<RoleScope>()
        var roles: [EligibleRole] = []
        for i in items {
            let scope = RoleScope.azureResource(scope: i.properties.scope, roleDefinitionId: i.properties.roleDefinitionId)
            guard seen.insert(scope).inserted else { continue }
            roles.append(EligibleRole(key: RoleKey(identityId: identity.id, tenantId: tenant.tenantId, scope: scope),
                                      displayName: i.properties.expandedProperties?.roleDefinition?.displayName ?? i.properties.roleDefinitionId,
                                      detail: Self.caption(i.properties.expandedProperties),
                                      source: .discovered, policy: .manualDefault))
        }
        return roles.sorted { ($0.displayName, $0.detail ?? "") < ($1.displayName, $1.detail ?? "") }
    }

    public func activeAssignments(identity: Identity, tenant: TenantContext) async throws -> [ActiveAssignment] {
        let instances = try await listAll(Instance.self, identity: identity, tenantId: tenant.tenantId,
                                          url: try armURL("providers/Microsoft.Authorization/roleAssignmentScheduleInstances", query: ["$filter": "asTarget()"]))
        let requests = try await listAll(Instance.self, identity: identity, tenantId: tenant.tenantId,
                                         url: try armURL("providers/Microsoft.Authorization/roleAssignmentScheduleRequests", query: ["$filter": "asTarget()"]))
        var result: [RoleKey: ActiveAssignment] = [:]
        for i in instances where i.properties.assignmentType == "Activated" {
            let key = RoleKey(identityId: identity.id, tenantId: tenant.tenantId, scope: .azureResource(scope: i.properties.scope, roleDefinitionId: i.properties.roleDefinitionId))
            result[key] = ActiveAssignment(roleKey: key, assignmentId: i.name, startDateTime: i.properties.startDateTime ?? .now,
                                           endDateTime: i.properties.endDateTime, status: .active)
        }
        for r in requests where Self.pendingStatuses.contains(r.properties.status ?? "") {
            let key = RoleKey(identityId: identity.id, tenantId: tenant.tenantId, scope: .azureResource(scope: r.properties.scope, roleDefinitionId: r.properties.roleDefinitionId))
            guard result[key] == nil else { continue }
            result[key] = ActiveAssignment(roleKey: key, assignmentId: r.name,
                                           startDateTime: r.properties.scheduleInfo?.startDateTime ?? r.properties.createdOn ?? .now,
                                           endDateTime: nil, status: .pendingApproval)
        }
        return Array(result.values)
    }

    // Task 4 replaces these.
    public func policy(for role: EligibleRole, identity: Identity) async throws -> RolePolicy {
        throw PIMError.unexpected(status: 501, body: "policy: Task 4")
    }
    public func activate(_ request: ActivationRequest, identity: Identity) async throws -> ActiveAssignment {
        throw PIMError.unexpected(status: 501, body: "activate: Task 4")
    }
    public func deactivate(_ assignment: ActiveAssignment, identity: Identity) async throws {
        throw PIMError.unexpected(status: 501, body: "deactivate: Task 4")
    }
    public func cancelPendingRequest(_ assignment: ActiveAssignment, identity: Identity) async throws {
        throw PIMError.unexpected(status: 501, body: "cancel: Task 4")
    }
}

File: windows/src/Elevate.Core/Providers/AzureResourceProvider.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using Elevate.Core.Auth;
using Elevate.Core.Models;
using Elevate.Core.Networking;
using Elevate.Core.Support;

namespace Elevate.Core.Providers;

/// <summary>PIM for Azure resource roles through Azure Resource Manager. Port of the Swift <c>AzureResourceProvider</c>.</summary>
public sealed class AzureResourceProvider : IPimProvider
{
    private readonly GraphTransport _transport;

    public AzureResourceProvider(IHttpClient http, ITokenProvider tokens)
        => _transport = new GraphTransport(http, tokens, GraphTransport.MapArmError);

    public RoleScopeKind Kind => RoleScopeKind.AzureResource;

    public IReadOnlyList<string> Scopes => Auth.Scopes.ArmAll;

    // MARK: Wire models

    private sealed record Named(string? DisplayName, string? Type, string? Id);

    private sealed record Expanded(Named? Scope, Named? RoleDefinition, Named? Principal);

    private sealed record Properties(
        string Scope,
        string RoleDefinitionId,
        string? PrincipalId,
        string? Status,
        string? AssignmentType,
        string? RoleEligibilityScheduleId,
        string? LinkedRoleEligibilityScheduleId,
        DateTimeOffset? StartDateTime,
        DateTimeOffset? EndDateTime,
        DateTimeOffset? CreatedOn,
        ScheduleInfo? ScheduleInfo,
        Expanded? ExpandedProperties,
        string? MemberType);

    private sealed record Instance(string Name, string Id, Properties Properties);

    /// <summary>One page of an ARM list; ARM names the continuation <c>nextLink</c>, not <c>@odata.nextLink</c>.</summary>
    private sealed record ArmPage<T>(IReadOnlyList<T>? Value, string? NextLink);

    private static readonly IReadOnlySet<string> PendingStatuses = new HashSet<string>(StringComparer.Ordinal)
    {
        "PendingApproval", "PendingAdminDecision", "PendingApprovalProvisioning",
    };

    /// <summary>An ARM URL for <paramref name="path"/> with <c>api-version</c> first and the rest of the query sorted.</summary>
    internal static Uri ArmUrl(string path, string apiVersion = "2020-10-01", IReadOnlyDictionary<string, string>? query = null)
    {
        var builder = new StringBuilder(GraphTransport.ArmBase)
            .Append('/').Append(path)
            .Append("?api-version=").Append(GraphTransport.PercentEncodeQuery(apiVersion));

        foreach (var (key, value) in (query ?? new Dictionary<string, string>()).OrderBy(p => p.Key, StringComparer.Ordinal))
        {
            builder.Append('&')
                .Append(GraphTransport.PercentEncodeQuery(key))
                .Append('=')
                .Append(GraphTransport.PercentEncodeQuery(value));
        }

        return Uri.TryCreate(builder.ToString(), UriKind.Absolute, out var url)
            ? url
            : throw new PimException(PimErrorKind.Unexpected, $"Bad ARM path {path}", 0);
    }

    /// <summary>A scope's path form for an ARM URL: no leading or trailing slash.</summary>
    private static string Trimmed(string scope) => scope.Trim('/');

    /// <summary>GETs every page of an ARM list, following <c>nextLink</c>.</summary>
    private async Task<IReadOnlyList<T>> ListAllAsync<T>(
        Identity identity, string tenantId, Uri url, CancellationToken ct)
    {
        Uri? next = url;
        var all = new List<T>();
        while (next is { } current)
        {
            var response = await _transport.GetAsync(identity, tenantId, current, Scopes, ct).ConfigureAwait(false);
            var page = JsonSerializer.Deserialize<ArmPage<T>>(response.Body, GraphJson.Options);
            if (page?.Value is { } items)
            {
                all.AddRange(items);
            }

            next = page?.NextLink is { } link && Uri.TryCreate(link, UriKind.Absolute, out var parsed) ? parsed : null;
        }

        return all;
    }

    /// <summary>The scope's display name and kind, e.g. "Pay-As-You-Go · subscription".</summary>
    private static string? Caption(Expanded? expanded)
    {
        if (expanded?.Scope is not { DisplayName: { } name })
        {
            return null;
        }

        var type = expanded.Scope.Type?.ToLowerInvariant() switch
        {
            "subscription" => "subscription",
            "resourcegroup" => "resource group",
            "managementgroup" => "management group",
            { } other => other,
            null => null,
        };

        return type is null ? name : string.Create(CultureInfo.InvariantCulture, $"{name} · {type}");
    }

    // MARK: Reads

    public async Task<IReadOnlyList<EligibleRole>> EligibleRolesAsync(
        Identity identity, TenantContext tenant, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        ArgumentNullException.ThrowIfNull(tenant);

        var items = await ListAllAsync<Instance>(identity, tenant.TenantId, EligibilityUrl(), ct).ConfigureAwait(false);

        var seen = new HashSet<RoleScope>();
        var roles = new List<EligibleRole>();
        foreach (var i in items)
        {
            var scope = new AzureResourceScope(i.Properties.Scope, i.Properties.RoleDefinitionId);
            if (!seen.Add(scope))
            {
                continue;
            }

            var viaGroup = string.Equals(i.Properties.MemberType, "Group", StringComparison.OrdinalIgnoreCase)
                ? i.Properties.ExpandedProperties?.Principal?.DisplayName ?? "group"
                : null;

            roles.Add(new EligibleRole(
                new RoleKey(identity.Id, tenant.TenantId, scope),
                i.Properties.ExpandedProperties?.RoleDefinition?.DisplayName ?? i.Properties.RoleDefinitionId,
                RoleSource.Discovered,
                RolePolicy.ManualDefault,
                Detail: Caption(i.Properties.ExpandedProperties),
                ViaGroup: viaGroup));
        }

        return
        [
            .. roles
                .OrderBy(r => r.DisplayName, StringComparer.Ordinal)
                .ThenBy(r => r.Detail ?? "", StringComparer.Ordinal),
        ];
    }

    public async Task<IReadOnlyList<ActiveAssignment>> ActiveAssignmentsAsync(
        Identity identity, TenantContext tenant, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(identity);
        ArgumentNullException.ThrowIfNull(tenant);

        var instances = await ListAllAsync<Instance>(
            identity, tenant.TenantId,
            ArmUrl("providers/Microsoft.Authorization/roleAssignmentScheduleInstances", query: AsTarget),
            ct).ConfigureAwait(false);

        var requests = await ListAllAsync<Instance>(
            identity, tenant.TenantId,
            ArmUrl("providers/Microsoft.Authorization/roleAssignmentScheduleRequests", query: AsTarget),
            ct).ConfigureAwait(false);

        var result = new Dictionary<RoleKey, ActiveAssignment>();
        foreach (var i in instances.Where(i => i.Properties.AssignmentType == "Activated"))
        {
            var key = Key(identity, tenant.TenantId, i.Properties);
            result[key] = new ActiveAssignment(
                key, i.Name, i.Properties.StartDateTime ?? DateTimeOffset.UtcNow,
                i.Properties.EndDateTime, AssignmentStatus.Active);
        }

        foreach (var r in requests.Where(r => PendingStatuses.Contains(r.Properties.Status ?? "")))
        {
            var key = Key(identity, tenant.TenantId, r.Properties);
            if (result.ContainsKey(key))
            {
                continue;
            }

            result[key] = new ActiveAssignment(
                key, r.Name,
                r.Properties.ScheduleInfo?.StartDateTime ?? r.Properties.CreatedOn ?? DateTimeOffset.UtcNow,
                null, AssignmentStatus.PendingApproval);
        }

        // The requests list is unfiltered, so a booked-ahead request is already in hand.
        foreach (var u in requests.Where(u => !ScheduleRules.IsSettledOrPending(u.Properties.Status)))
        {
            if (u.Properties.ScheduleInfo?.StartDateTime is not { } start || !ScheduleRules.IsFuture(start))
            {
                continue;
            }

            var key = Key(identity, tenant.TenantId, u.Properties);
            if (result.ContainsKey(key))
            {
                continue;
            }

            var end = ScheduleRules.End(
                u.Properties.ScheduleInfo?.Expiration?.EndDateTime,
                u.Properties.ScheduleInfo?.Expiration?.Duration,
                start);
            result[key] = new ActiveAssignment(key, u.Name, start, end, AssignmentStatus.Scheduled);
        }

        return [.. result.Values];
    }

    private static readonly IReadOnlyDictionary<string, string> AsTarget =
        new Dictionary<string, string>(StringComparer.Ordinal) { ["$filter"] = "asTarget()" };

    private static Uri EligibilityUrl() =>
        ArmUrl("providers/Microsoft.Authorization/roleEligibilityScheduleInstances", query: AsTarget);

    private static RoleKey Key(Identity identity, string tenantId, Properties properties) =>
        new(identity.Id, tenantId, new AzureResourceScope(properties.Scope, properties.RoleDefinitionId));

    // MARK: Policy

    private sealed record PolicyProperties(string? RoleDefinitionId, IReadOnlyList<PolicyRules.Rule>? EffectiveRules);

    private sealed record PolicyAssignment(string Name, PolicyProperties Properties);

    public async Task<RolePolicy> PolicyAsync(EligibleRole role, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(role);

        if (role.Key.Scope is not AzureResourceScope scope)
        {
            throw new PimException(PimErrorKind.NotEligible);
        }

        var url = ArmUrl(Trimmed(scope.Scope) + "/providers/Microsoft.Authorization/roleManagementPolicyAssignments");
        var assignments = await ListAllAsync<PolicyAssignment>(identity, role.Key.TenantId, url, ct).ConfigureAwait(false);

        var match = assignments.FirstOrDefault(a =>
            string.Equals(a.Properties.RoleDefinitionId, scope.RoleDefinitionId, StringComparison.OrdinalIgnoreCase));
        return match?.Properties.EffectiveRules is { } rules ? PolicyRules.Apply(rules) : RolePolicy.ManualDefault;
    }

    // MARK: Activation

    private sealed record RoleDefinition(string Id, RoleDefinition.Props Properties)
    {
        public sealed record Props(string? RoleName);
    }

    /// <summary>Manual roles carry a role <em>name</em>; ARM wants the definition id at that scope.</summary>
    internal async Task<string> ResolveRoleDefinitionIdAsync(
        string nameOrId, string scope, Identity identity, string tenantId, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(nameOrId);

        if (nameOrId.Contains('/', StringComparison.Ordinal))
        {
            return nameOrId;
        }

        var url = ArmUrl(
            Trimmed(scope) + "/providers/Microsoft.Authorization/roleDefinitions",
            "2022-04-01",
            new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["$filter"] = $"roleName eq '{GraphTransport.OdataEscaped(nameOrId)}'",
            });

        var definitions = await ListAllAsync<RoleDefinition>(identity, tenantId, url, ct).ConfigureAwait(false);
        return definitions.Count > 0 ? definitions[0].Id : throw new PimException(PimErrorKind.NotEligible);
    }

    /// <summary>Finds the caller's eligibility for a scope + role; ARM needs its principal id and schedule name to activate.</summary>
    private async Task<(string PrincipalId, string ScheduleName)> EligibilityAsync(
        string scope, string roleDefinitionId, Identity identity, string tenantId, CancellationToken ct)
    {
        var items = await ListAllAsync<Instance>(identity, tenantId, EligibilityUrl(), ct).ConfigureAwait(false);
        var match = items.FirstOrDefault(i =>
            string.Equals(i.Properties.Scope, scope, StringComparison.OrdinalIgnoreCase)
            && string.Equals(i.Properties.RoleDefinitionId, roleDefinitionId, StringComparison.OrdinalIgnoreCase));

        if (match?.Properties.PrincipalId is not { } principal
            || match.Properties.RoleEligibilityScheduleId is not { } schedule)
        {
            throw new PimException(PimErrorKind.NotEligible);
        }

        // Swift: components(separatedBy: "/").last ?? schedule — a trailing slash yields "".
        return (principal, schedule[(schedule.LastIndexOf('/') + 1)..]);
    }

    /// <summary>
    /// The principal to put in a Self* request: always the caller. An eligibility inherited
    /// through a group carries the <em>group's</em> principal id, and ARM refuses a request naming it
    /// ("The requestor … does not have permissions for this request"). The caller's object id
    /// in this tenant comes from the ARM token; an opaque token falls back to the eligibility's id.
    /// </summary>
    private async Task<string> RequestPrincipalIdAsync(
        string eligibilityPrincipalId, Identity identity, string tenantId, CancellationToken ct)
    {
        try
        {
            var token = await _transport.Tokens.AccessTokenAsync(identity, tenantId, Scopes, ct).ConfigureAwait(false);
            if (AccessTokenClaims.ObjectId(token) is { } oid)
            {
                return oid;
            }
        }
        catch (OperationCanceledException) when (ct.IsCancellationRequested)
        {
            throw;
        }
        catch (PimException)
        {
            // Fall through to the eligibility's own principal, like the Swift `try?`.
        }

        return eligibilityPrincipalId;
    }

    private static Uri RequestUrl(string scope) =>
        ArmUrl(Trimmed(scope) + "/providers/Microsoft.Authorization/roleAssignmentScheduleRequests/"
            + Guid.NewGuid().ToString("D", CultureInfo.InvariantCulture).ToLowerInvariant());

    private static byte[] Encode(JsonObject properties) =>
        Encoding.UTF8.GetBytes(new JsonObject { ["properties"] = properties }.ToJsonString());

    public async Task<ActiveAssignment> ActivateAsync(
        ActivationRequest request, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(request);

        if (request.RoleKey.Scope is not AzureResourceScope scope)
        {
            throw new PimException(PimErrorKind.NotEligible);
        }

        var tenantId = request.RoleKey.TenantId;
        var roleDefinitionId = await ResolveRoleDefinitionIdAsync(scope.RoleDefinitionId, scope.Scope, identity, tenantId, ct)
            .ConfigureAwait(false);
        var eligibility = await EligibilityAsync(scope.Scope, roleDefinitionId, identity, tenantId, ct).ConfigureAwait(false);
        var principalId = await RequestPrincipalIdAsync(eligibility.PrincipalId, identity, tenantId, ct).ConfigureAwait(false);

        var props = new JsonObject
        {
            ["principalId"] = principalId,
            ["roleDefinitionId"] = roleDefinitionId,
            ["requestType"] = "SelfActivate",
            ["linkedRoleEligibilityScheduleId"] = eligibility.ScheduleName,
            ["justification"] = request.Justification,
            ["scheduleInfo"] = new JsonObject
            {
                ["startDateTime"] = GraphJson.EncoderDateString(request.StartDateTime ?? DateTimeOffset.UtcNow),
                ["expiration"] = new JsonObject
                {
                    ["type"] = "AfterDuration",
                    ["duration"] = Iso8601Duration.Format(request.Duration),
                },
            },
        };

        if (request.Ticket is { } ticket)
        {
            props["ticketInfo"] = new JsonObject
            {
                ["ticketNumber"] = ticket.Number,
                ["ticketSystem"] = ticket.System,
            };
        }

        var response = await _transport
            .PutAsync(identity, tenantId, RequestUrl(scope.Scope), Scopes, Encode(props), ct).ConfigureAwait(false);
        var created = JsonSerializer.Deserialize<Instance>(response.Body, GraphJson.Options)
            ?? throw new PimException(PimErrorKind.Unexpected, "Empty response body");

        var start = ScheduleRules.Effective(created.Properties.ScheduleInfo?.StartDateTime, request.StartDateTime);
        var end = ScheduleRules.End(
            created.Properties.ScheduleInfo?.Expiration?.EndDateTime,
            created.Properties.ScheduleInfo?.Expiration?.Duration,
            start, request.Duration);

        var raw = created.Properties.Status ?? "Provisioned";
        var reported = raw switch
        {
            "PendingApproval" or "PendingAdminDecision" or "PendingApprovalProvisioning" => AssignmentStatus.PendingApproval,
            "PendingProvisioning" or "PendingScheduleCreation" or "ScheduleCreated" or "Accepted"
                or "PendingEvaluation" or "ProvisioningStarted" or "PendingExternalProvisioning"
                => AssignmentStatus.PendingProvisioning,
            "Denied" or "Failed" or "Canceled" or "Revoked" or "TimedOut" or "Invalid" or "AdminDenied"
                or "FailedAsResourceIsLocked" => AssignmentStatus.Failed(raw),
            _ => AssignmentStatus.Active,
        };

        // A future start only masks an outcome that would otherwise read as active; pending/failed still win.
        var status = reported == AssignmentStatus.Active && ScheduleRules.IsFuture(start)
            ? AssignmentStatus.Scheduled
            : reported;

        // A manual role is keyed by role name; key the assignment by the id ARM resolved it to.
        var resolvedKey = new RoleKey(
            request.RoleKey.IdentityId, tenantId, new AzureResourceScope(scope.Scope, roleDefinitionId));

        return new ActiveAssignment(
            resolvedKey, created.Name, start,
            status == AssignmentStatus.Active || status == AssignmentStatus.Scheduled ? end : null,
            status);
    }

    public async Task DeactivateAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(assignment);

        if (assignment.RoleKey.Scope is not AzureResourceScope scope)
        {
            throw new PimException(PimErrorKind.NotEligible);
        }

        var tenantId = assignment.RoleKey.TenantId;
        var roleDefinitionId = await ResolveRoleDefinitionIdAsync(scope.RoleDefinitionId, scope.Scope, identity, tenantId, ct)
            .ConfigureAwait(false);
        var eligibility = await EligibilityAsync(scope.Scope, roleDefinitionId, identity, tenantId, ct).ConfigureAwait(false);
        var principalId = await RequestPrincipalIdAsync(eligibility.PrincipalId, identity, tenantId, ct).ConfigureAwait(false);

        var props = new JsonObject
        {
            ["principalId"] = principalId,
            ["roleDefinitionId"] = roleDefinitionId,
            ["requestType"] = "SelfDeactivate",
            ["linkedRoleEligibilityScheduleId"] = eligibility.ScheduleName,
        };

        await _transport.PutAsync(identity, tenantId, RequestUrl(scope.Scope), Scopes, Encode(props), ct)
            .ConfigureAwait(false);
    }

    public async Task CancelPendingRequestAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(assignment);

        if (assignment.RoleKey.Scope is not AzureResourceScope scope || assignment.AssignmentId is not { } name)
        {
            throw new PimException(PimErrorKind.NotEligible);
        }

        var url = ArmUrl(
            Trimmed(scope.Scope) + $"/providers/Microsoft.Authorization/roleAssignmentScheduleRequests/{name}/cancel");
        await _transport.PostAsync(identity, assignment.RoleKey.TenantId, url, Scopes, [], ct).ConfigureAwait(false);
    }
}
